configs/prediction/default.yaml

```diff
@@ -12,3 +12,13 @@ dataset:
 
 device: cuda
 out_key: inpainted
+
+refine: False # refiner will only run if this is True
+refiner:
+  gpu_ids: 0,1 # the GPU ids of the machine to use. If only single GPU, use: "0,"
```

Comment:
I'd suggest using only 0 by default - or even introduce "None" default (so refiner would rely on the parent `device` setting). That would make this work by default in more environments without any modifications by default.

Reply:
Actually the refiner needs around 24GB GPU to process 1.8 megapixel images (~1200x1500). Since most people have two 12GB GPUs instead, we decided to split the model onto two GPUs, that's why the default config setting. 
Do you suggest to still make it None by default?

Reply:
torch.cuda.amp isn't working because pytorch doesn't seem to support `Half` dtype for `torch.fft.rfftn`. PFA link to the relevant issues in Pytorch repo:

https://github.com/pytorch/pytorch/issues/70664
https://github.com/pytorch/pytorch/issues/71680

Reply:
> I also manually verified the requires_grad for all the params of the model

Great, thank you!

> torch.cuda.amp isn't working because pytorch doesn't seem to support Half 

Sure, I've forgot that I've already tried half and failed because of that... We could wrap rfftn/irfftn with conversion to and from .float(), but I'm not sure there wouldn't be other issues..

Reply:
Hello @windj007, sorry for coming back after 2 months! We picked up the experiments, our findings:

1) We were able to perform the optimization in mixed precision. I haven't benchmarked it quantitatively, but qualitative results look good. However, for 1024x1024 images, it only reduces the memory from `21-22GB -> 17-18GB`, so it is still not sufficient to fit on a single 12GB GPU
2) We also tried to play with checkpointing. Performing it naively throws `RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn`, which we bypass by setting [`use_reentrant=False`](https://pytorch.org/docs/stable/checkpoint.html). However, this setting has some memory leak problem, which causes the GPU consumption to increase at each training loop, eventually leading to OOM error. We plan to raise this issue on the Pytorch repo.

Reply:
Can you please share your code for mixed-precision?

Reply:
Sure, it's in `amp_float16` branch of our fork:

[https://github.com/geomagical/lama-with-refiner/tree/amp_float16](https://github.com/geomagical/lama-with-refiner/tree/amp_float16)

You can get this code by:
```
git clone git@github.com:geomagical/lama-with-refiner.git
git checkout amp_float16
```

Also, I've changed the config file of the refiner to run on a single GPU. But yeah feel free to play around with config parameters or anything :)

Reply:
Thank you so much 🙏